sklearn/metrics/pairwise.py

```diff
@@ -401,14 +407,26 @@ def _euclidean_distances(X, Y, X_norm_squared=None, Y_norm_squared=None, squared
         distances = -2 * safe_sparse_dot(X, Y.T, dense_output=True)
         distances += XX
         distances += YY
-    np.maximum(distances, 0, out=distances)
+
+    xp_zero = xp.asarray(0, device=device_, dtype=distances.dtype)
+    if _xp_method_has_out(xp_method=xp.maximum):
+        xp.maximum(distances, xp_zero, out=distances)
+    else:
+        distances = xp.maximum(distances, xp_zero)
 
     # Ensure that distances between vectors and themselves are set to 0.0.
     # This may not be the case due to floating point rounding errors.
     if X is Y:
-        np.fill_diagonal(distances, 0)
+        _add_to_diagonal(distances, 0, xp=xp, add_value=False)
```

Comment:
This function could be renamed to `_fill_or_add_to_diagonal` to improve readability.